File: lesson1-iterations/BinaryGap.swift
/*
    Codility Lesson 1 - BinaryGap
    : Find longest sequence of zeros in binary representation of an integer.

    - Level of Difficulty: Painless
    - Task Score: 100%
    - Correntness: 100%
*/

import Foundation
import Glibc

public func solution(_ N: Int) -> Int {
    let binaryArray = makeBinaryArray(N)
    var gaps = [Int]()
    var gap = 0
    
    for (index, binary) in binaryArray.enumerated() {
        if binary == 0 {
            guard index != binaryArray.endIndex else { break }
            gap += 1
        } else {
            gaps.append(gap)
            gap = 0
        }
    }
    
    return gaps.max()!
}

private func makeBinaryArray(_ N: Int) -> [Int] {
    var integer = N
    var array = [Int]()
    
    while integer != 0 {
        array.insert(integer % 2, at: 0)
        integer /= 2
    }
    
    return array
}

File: Codility/BinaryGap.java
/*
    BinaryGap
    : Find longest sequence of zeros in binary representation of an integer.

    - Difficulty Level: PAINLESS

    - Task Score: 100%
    - Correctness: 100%
 */

import java.util.*;

class Solution {
    public int solution(int N) {
        ArrayList<Integer> binaryList = makeBinaryList(N);
        int maxGap = getMaximumGap(binaryList);

        return maxGap;
    }

    private static ArrayList<Integer> makeBinaryList(int N) {
        ArrayList<Integer> binaryList = new ArrayList<>();
        int currentN = N;

        while (currentN > 0) {
            binaryList.add(currentN % 2);
            currentN /= 2;
        }

        return binaryList;
    }

    private static int getMaximumGap(ArrayList<Integer> list) {
        int maxGap = 0;
        int tempGap = 0;

        for (int i = list.size() - 2; i >= 0; i--) {
            Integer binary = list.get(i);

            if (binary == 1) {
                maxGap = (tempGap > maxGap) ? tempGap : maxGap;
                tempGap = 0;
            } else {
                tempGap += 1;
            }
        }

        return maxGap;
    }
}